### PR TITLE
Handle block change

### PR DESCRIPTION
**Pull request by olafurj**
a few fixes and additions to make it work

### CODE REVIEW

**Comment by nickelpro** [comment]
The bug in the bitwise decoding is my fault, basically a transcription error.

Normally what you're doing here, bitshifting a signed int, is implementation-defined behavior and wouldn't work the way you're expecting it to on all compilers. Thankfully this became defined behavior in C++20 which is the standard we target, so we can handle it this way and simplify the code. Well done, thanks again.

File: src/smpmap.cpp [hunk 1]
```diff
@@ -135,6 +135,19 @@ void SMPMap::update(const mcd::ClientboundMultiBlockChange& packet) {
       packet.chunkCoordinates.y, packet.records);
 }
 
+void SMPMap::update(const mcd::ClientboundBlockChange& packet) {
+  std::unique_lock lock(mutex);
+  if (chunks[{packet.location.x >> 4, packet.location.z >> 4}].sections[packet.location.y >> 4])
```
Comment (nickelpro) [inline src/smpmap.cpp:140]: What happens if this is getting placed into a chunk section that hasn't been created yet? It's good to check, but the chunk section needs to be `.emplaced()` if it doesn't already exist. Ideally this should happen inside the relevant `ChunkColumn`, not `SMPMap`.

File: src/smpmap.cpp [hunk 2]
```diff
@@ -135,6 +135,19 @@ void SMPMap::update(const mcd::ClientboundMultiBlockChange& packet) {
...
+  {
+    std::vector<std::int64_t> records;
+    std::int64_t block=packet.type << 12 | ((packet.location.x & 0xf) << 8 | (packet.location.y & 0xf) << 4 | (packet.location.z & 0xf));
+    records.push_back(block);
+    chunks[{packet.location.x >> 4, packet.location.z >> 4}].update(
```
Comment (nickelpro) [inline src/smpmap.cpp:145]: Just create a new update method for single block updates, don't hijack the machinery for multi block updates. Building this vector isn't necessary

File: src/smpmap.cpp [hunk 3]
```diff
@@ -88,9 +88,9 @@ void ChunkColumn::update(std::uint8_t sec_coord,
     const std::vector<std::int64_t>& records) {
   auto& section = sections[sec_coord];
   for(auto el : records) {
-    std::uint8_t y = el&0xF;
-    std::uint8_t z = (el>>=4)&0xF;
-    std::uint8_t x = (el>>=4)&0xF;
+    std::int64_t z = el&0xF;
+    std::int64_t y = (el>>=4)&0xF;
+    std::int64_t x = (el>>=4)&0xF;
```
Comment (nickelpro) [inline src/smpmap.cpp:93]: Were these broken? The max value of chunk relative coordinates is 255

**Review by nickelpro — CHANGES_REQUESTED** [review]

**Comment by nickelpro** [comment]
Oooh I see what the problem in my dec_position is, it's from pre-1.14, back when the position bitfield layout was different. Might want to look into generating that from minecraft-data instead of hand coding it...

**Comment by nickelpro** [comment]
This should all be fixed now, closing in favor of https://github.com/SpockBotMC/RikerBot/commit/0f0e1e46f56cb03cc2e570439f76b3eb05447758

Thank you Ólafur!

**Comment by olafurj** [comment]
thats why I check if section exists before, nothing happens if it does notSent from my Galaxy
-------- Original message --------From: Vito Gamberini <notifications@github.com> Date: 11/30/20  05:23  (GMT+00:00) To: SpockBotMC/RikerBot <RikerBot@noreply.github.com> Cc: Olafur Jon Bjornsson <ojb@simnet.is>, Author <author@noreply.github.com> Subject: Re: [SpockBotMC/RikerBot] Handle block change (#15) 
@nickelpro requested changes on this pull request.



In src/smpmap.cpp:
> @@ -135,6 +135,19 @@ void SMPMap::update(const mcd::ClientboundMultiBlockChange& packet) {
       packet.chunkCoordinates.y, packet.records);
 }
 
+void SMPMap::update(const mcd::ClientboundBlockChange& packet) {
+  std::unique_lock lock(mutex);
+  if (chunks[{packet.location.x >> 4, packet.location.z >> 4}].sections[packet.location.y >> 4])

What happens if this is getting placed into a chunk section that hasn't been created yet? It's good to check, but the chunk section needs to be .emplaced() if it doesn't already exist. Ideally this should happen inside the relevant ChunkColumn, not SMPMap.



In src/smpmap.cpp:
> @@ -135,6 +135,19 @@ void SMPMap::update(const mcd::ClientboundMultiBlockChange& packet) {
       packet.chunkCoordinates.y, packet.records);
 }
 
+void SMPMap::update(const mcd::ClientboundBlockChange& packet) {
+  std::unique_lock lock(mutex);
+  if (chunks[{packet.location.x >> 4, packet.location.z >> 4}].sections[packet.location.y >> 4])
+  {
+    std::vector<std::int64_t> records;
+    std::int64_t block=packet.type << 12 | ((packet.location.x & 0xf) << 8 | (packet.location.y & 0xf) << 4 | (packet.location.z & 0xf));
+    records.push_back(block);
+    chunks[{packet.location.x >> 4, packet.location.z >> 4}].update(

Just create a new update method for single block updates, don't hijack the machinery for multi block updates. Building this vector isn't necessary



In src/smpmap.cpp:
> @@ -88,9 +88,9 @@ void ChunkColumn::update(std::uint8_t sec_coord,
     const std::vector<std::int64_t>& records) {
   auto& section = sections[sec_coord];
   for(auto el : records) {
-    std::uint8_t y = el&0xF;
-    std::uint8_t z = (el>>=4)&0xF;
-    std::uint8_t x = (el>>=4)&0xF;
+    std::int64_t z = el&0xF;
+    std::int64_t y = (el>>=4)&0xF;
+    std::int64_t x = (el>>=4)&0xF;

Were these broken? The max value of chunk relative coordinates is 255

File: src/smpmap.cpp (hunk 3 above)
Comment (olafurj) [reply src/smpmap.cpp:93]: if I left them as they were, they did not have valid values, this was a quick fix that solved it, plus the order was wrong according to https://wiki.vg/Protocol#Multi_Block_Change.
You can propably do some casting tricks to make it work with 8 bits

**Comment by olafurj** [comment]
stil a little shy to change the code to much, but you are right new method is much cleaner :-)

**Pull request closed without merging**